examples/decorations/graphs.jl

```diff
@@ -0,0 +1,401 @@
+
+remove decorations of a span of decorated morphisms
+"""
+function undecorate(s::Span{T,T}) where T <: Decorated
+    return Span(undecorate(left(s)), undecorate(right(s)))
+end
+
+# +
+struct DoublePushout{S<:AbstractSpan, T<:NTuple{3,AbstractMorph}}
+    rule::S
+    morphs::T
+    application::S
+end
+
+# TODO DPO CONSTRUCTOR TO SOLVE UNKNOWN DOUBLEPUSHOUT
+
+# +
+"""    AbstractCospan
+
+an abstract type for representing cospans. The essential API for subtypes of AbstractCospan are
+
+1. left(c)::M
+2. right(c)::M
+3. pullback(c)::S
+
+where M is the type of morphisms in the cospan, and S is the type of span that solves the
+pullback defined by c. See Cospan for an example.
+"""
+abstract type AbstractCospan end
+
+leftob(c::AbstractCospan) = codom(left(c))
+
+rightob(c::AbstractCospan) = codom(right(c))
+
+function apexob(c::AbstractCospan)
+    a = dom(left(c))
+    b = dom(right(c))
+    a == b || error("Inconsistent cospan")
+    return a
+end
+
+# +
+"""    Cospan{F,G} <: AbstractCospan
+
+a general cospan type where types F and G are types of morphisms in the cospan
+"""
+struct Cospan{F,G} <: AbstractCospan
+    f::F
+    g::G
+end
+
+function left(c::Cospan)
+    return c.f
+end
+
+function right(c::Cospan)
+    return c.g
+end
+
+"""    undecorate(s::Copan{T,T}) where T <: Decorated
+
+remove decorations of a cospan of decorated morphisms
+"""
+function undecorate(c::Cospan{T,T}) where T <: Decorated
+    return Cospan(undecorate(left(c)), undecorate(right(c)))
+end
+
+# +
+"""    pushout(s::Span{T,T}) where T <: FinSetMorph
+
+treat f,g as a span and compute the pushout that is, the cospan of f=(f⊔g) and g=(a⊔b)
+"""
+function pushout(s::Span{T,T}) where T <: FinSetMorph
+    f_dict = Dict(a=>i for (i, a) in enumerate(left(s).fun))
+    g′ = map(n->n in keys(f_dict) ? func(right(s))(f_dict[n]) : n+length(rightob(s)), leftob(s))
+    
+    g_dict = Dict(a=>i for (i, a) in enumerate(right(s).fun))
+    f′ = map(n->n in keys(g_dict) ? g′[func(left(s))(g_dict[n])] : n, rightob(s))
+
+    u = union(f′, g′)
+    u_dict = Dict(a=>i for (i, a) in enumerate(u))
+    f′ = FinSetMorph(1:length(u), map(n->u_dict[n], f′))
+    g′ = FinSetMorph(1:length(u), map(n->u_dict[n], g′))
+
+    return Cospan(f′, g′)
+end
+
+"""    pushout(s::Span{T, T}) where T <: Decorated
+
+treat f,g as a decorated span and compute the pushout that is, the cospan of f=(f⊔g) and g=(a⊔b),
+with the decoration of (f⊔g)(d)
+"""
+function pushout(s::Span{T, T}) where T <: Decorated
+    cs = pushout(undecorate(s))
+    D = decoration(left(s)) ⊔ decoration(right(s))
+    return Decorated(cs, (right(cs) ⊔ left(cs))(D))
+end
+# -
+
+# explicit pushout definition for testing and verification of span implementation
+function pushout(a::AbstractGraph, b::AbstractGraph, f::AbstractVector{Int}, g::AbstractVector{Int})
+    l = f ⊔ g
+    G = a ⊔ b
+    map(edges(G)) do e
+        s,t = e.src, e.dst
+        return Edge(l[s], l[t])
+    end |> Graph
+end
+
```

Comment:
everything above this line should be migrated to `src`, everything after should be copied and adapted to `test`.